tardis/plasma/properties/util/macro_atom.pyx

```diff
@@ -0,0 +1,99 @@
+# module for fast macro_atom calculations
+
+# cython: profile=False
+# cython: boundscheck=False
+# cython: cdivision=True
+# cython: wraparound=False
+
+import numpy as np
+
+cimport numpy as np
+
+ctypedef np.int64_t int_type_t
+
+from astropy import constants
+
+from cython.parallel import prange
+
+
+cdef extern from "math.h":
+    double exp(double)
+
+
+cdef double h_cgs = constants.h.cgs.value
+cdef double c = constants.c.cgs.value
+cdef double kb = constants.k_B.cgs.value
+cdef double inv_c2 = 1 / (c ** 2)
+
+
+def calculate_beta_sobolev(np.ndarray[double, ndim=1] tau_sobolevs, np.ndarray[double, ndim=1] beta_sobolevs):
+    cdef double beta_sobolev
+    cdef double tau_sobolev
+    cdef int i
+
+    for i in range(len(tau_sobolevs)):
+        tau_sobolev = tau_sobolevs[i]
+
+        if tau_sobolev > 1e3:
+            beta_sobolev = 1 / tau_sobolev
+        elif tau_sobolev < 1e-4:
+            beta_sobolev = 1 - 0.5 * tau_sobolev
+        else:
+            beta_sobolev = (1 - exp(-tau_sobolev)) / tau_sobolev
+        beta_sobolevs[i] = beta_sobolev
+
+def normalize_transition_probabilities(double [:, ::1] p_transition,
+                                       int_type_t [:] reference_levels):
+    cdef int i, j, k
+    cdef np.ndarray[double, ndim=1] norm_factor = np.zeros(p_transition.shape[1])
+    cdef int start_id = 0
+    cdef  end_id = 0
+
+    for i in range(len(reference_levels) - 1):
+        norm_factor[:] = 0.0
+        for j in range(reference_levels[i], reference_levels[i + 1]):
+            for k in range(p_transition.shape[1]):
+                norm_factor[k] += p_transition[j, k]
+        for j in range(reference_levels[i], reference_levels[i + 1]):
+            for k in range(0, p_transition.shape[1]):
+                if norm_factor[k] == 0.0:
+                    continue
+
+                p_transition[j, k] /= norm_factor[k]
+
+
+def calculate_transition_probabilities(
```

Comment:
On 08/25/15 18:25, Wolfgang Kerzendorf wrote:
>> +    cdef  end_id = 0
>> +
>> +    for i in range(len(reference_levels) - 1):
>> +        norm_factor[:] = 0.0
>> +        for j in range(reference_levels[i], reference_levels[i + 1]):
>> +            for k in range(p_transition.shape[1]):
>> +                norm_factor[k] += p_transition[j, k]
>> +        for j in range(reference_levels[i], reference_levels[i + 1]):
>> +            for k in range(0, p_transition.shape[1]):
>> +                if norm_factor[k] == 0.0:
>> +                    continue
>> +
>> +                p_transition[j, k] /= norm_factor[k]
>> +
>> +
>> +def calculate_transition_probabilities(
> 
> @mreinecke this is the beast. You could just write it in C and maybe with openmp it would be faster. I'm wondering if memory bandwidth is an issue here. 

It may be possible to optimize this a bit, but there are a few issues to
consider first:

- how large are the individual loop iteration counts (roughly, order of
magnitude is fine)? This determines the best arrangement of the nested
loops.

- Is it known how the array p_transition is laid out in memory on the
Python side? We won't gain any performace by tweaking the routine as
long as the glue code between Python and C has to do things like array
copying and maybe even transposition. This is because this function is
absolutely dominated by memory accesses; the cost of arithmetic is
negligible in comparison.

- related to the point above: does "p_transition[j, k]" in Cython mean
the same as "p_transition[j][k]" in C? More specifically, will
p_transition[j][k] and p_transition[j][k+1] be neighbours in memory?


Overall it would be beneficial if "p_transition" could be stored in a
way that elements with j and j+1 are neighbouring in memory, but I'm not
sure if the memory layout is constrained somehow by other considerations.

Cheers,
  Martin